RoboJackets/igvc-software
# Feat/swerve wheel odometry

# Description

Swerve Drive Wheel Odometry

This PR does the following:
- Implements wheel odometry using Instantaneous Centre of Rotation (ICR) 
- Fixes parameter issue with swerve drive controller node
- Correct wheel radius in swerve gazebo control node
- Fixed frame id in differential drive odometer

Fixes #868 

# Testing steps

## Test Case 1
1. Run `roslaunch igvc_gazebo ramp_lane.launch model:=1`
2. Run `roslaunch igvc_navigation swerve_drive.launch`
3. Run `roslaunch igvc_navigation swerve_wheel_odometry.launch `
4. Open rviz and visualize the `/ground_truth` topic and `/wheel_odometry` topic
5. Run
```
rostopic pub /cmd_vel geometry_msgs/Twist "linear:
  x: 1.0
  y: 0.0
  z: 0.0
angular:
  x: 0.0
  y: 0.0
  z: 0.0"
```

Expectation: Wheel odometry estimate should match the ground truth in rviz. Swervi should go directly forward.

## Test Case 2
1. Run `roslaunch igvc_gazebo ramp_lane.launch model:=1`
2. Run `roslaunch igvc_navigation swerve_drive.launch`
3. Run `roslaunch igvc_navigation swerve_wheel_odometry.launch `
4. Open rviz and visualize the `/ground_truth` topic and `/wheel_odometry` topic
5. Run
```
rostopic pub /cmd_vel geometry_msgs/Twist "linear:
  x: 0.0
  y: 1.0
  z: 0.0
angular:
  x: 0.0
  y: 0.0
  z: 0.0"
```

Expectation: Wheel odometry estimate should match the ground truth in rviz. Swervi should go to the left.

## Test Case 3
1. Run `roslaunch igvc_gazebo ramp_lane.launch model:=1`
2. Run `roslaunch igvc_navigation swerve_drive.launch`
3. Run `roslaunch igvc_navigation swerve_wheel_odometry.launch `
4. Open rviz and visualize the `/ground_truth` topic and `/wheel_odometry` topic
5. Run
```
rostopic pub /cmd_vel geometry_msgs/Twist "linear:
  x: 0.0
  y: 0.0
  z: 0.0
angular:
  x: 0.0
  y: 0.0
  z: 1.0"
```

Expectation: Wheel odometry estimate should match the ground truth in rviz. Swervi should rotate counter clockwise.

## Test Case 4
1. Run `roslaunch igvc_gazebo ramp_lane.launch model:=1`
2. Run `roslaunch igvc_navigation swerve_drive.launch`
3. Run `roslaunch igvc_navigation swerve_wheel_odometry.launch `
4. Open rviz and visualize the `/ground_truth` topic and `/wheel_odometry` topic
5. Run
```
rostopic pub /cmd_vel geometry_msgs/Twist "linear:
  x: -1.0
  y: 0.0
  z: 0.0
angular:
  x: 0.0
  y: 0.0
  z: 0.5"
```

Expectation: Wheel odometry estimate should match the ground truth in rviz. Swervi should travel in a circular motion backwards.

# Self Checklist
